### ISSUE TITLE
fix automatic build on Travis CI

### ISSUE THREAD

**Issue by achimmihca** [UltraStar-Deluxe/Play#189]
### Actual behaviour
The Travis CI build is always failing since the upgrade to Unity 2020

### Expected behaviour
The build should work for Windows, Linux, Mac.

Im am interested in fixing this.

- Is there documentation how the build works?
- Is it a matter of changing tools/travis?
- Where is the resulting executable of a build?
    - Is it a private server location?
        - If yes, could we find an alternative for UltraStar-Deluxe organisation, e.g., some web space (Goggle Drive, Microsoft OneDrive, Dropbox, etc.) 
- Would it be possible to also build Android automatically?

**Comment by achimmihca** [comment]
Documentation should be added here: https://github.com/UltraStar-Deluxe/Play/wiki/Continuous-Integration-Builds

**Comment by achimmihca** [comment]
@basisbit do you have some inputs for me to get started?

**Comment by basisbit** [comment]
TravisCI is used as buildbot. The configration can be found here: https://github.com/UltraStar-Deluxe/Play/blob/master/.travis.yml
For any event targeting the master branch, it will run the before_install steps, then the job task. For every build target / job it will build a docker container, install unity in there and try to build the unity project.
The result will be uploaded to a Amazon AWS s3 bucket which so far never reached its limit of free storage space or free download traffic, so no need to worry about that. Builds and other uploaded stuff are automatically deleted after two months I think. The download URL for the builds will be logged at the end of each build.

TravisCI build only uploads build results of the master branch to the s3 bucket, not any results of a pull request - otherwise it would be very easy to get the s3 secret as well as other secrets which can be set in the project page on TravisCI.

When upgrading to a newer major Unity version, then someone will have to run the unity based build in a local docker container and there go through the manual steps to active the unity install. Once that is done, before stopping the unity container, copy the unity license file and replace the file at https://github.com/UltraStar-Deluxe/Play/blob/master/tools/travis/activat.ion

**Comment by achimmihca** [comment]
Ok, I understand now how the build works and was able to fix it on my fork. I will document my findings soon.
> The past year, unity always changed the command a bit for how to activate / enable / local the install on a headless system

Indeed, I also ran into an issue. Luckily I found this [blog post](https://johnaustin.io/articles/2020/running-unity-20201-in-docker). The virtual X Server is (a) not needed anymore and (b) there is a segmentation fault when it is used.
Anyway, it worked after changing this.

However, I am confused because it seems that you intent to keep the license file secret. But "activat.ion" **_is_** the license file and it has already been committed. Username and password are not needed when the license file is present. So if the license file is in the repo, then you do not need to specify any secret on Travis.

Take a look at .travil.yml. It contains `export UNITY_LICENSE_CONTENT='cat tools/travis/activat.ion'`. This overwrites the secret env variable that you have set in Travis.
What happens at the moment is:

- write content of activat.ion to UNITY_LICENSE_CONTENT
- write content of UNITY_LICENSE_CONTENT to /root/.local/share/unity3d/Unity/Unity_lic.ulf

instead you can just write content of "activat.ion" to /root/.local/share/unity3d/Unity/Unity_lic.ulf

**Question is**: should the license file be in the repository in the first place? As you said, you tried some stuff before finding a working config. I can imagine that this was not intended.

If it should not be in the repo, then UNITY_LICENSE_CONTENT is the only env variable that needs to be set in Travis. And `export UNITY_LICENSE_CONTENT='cat tools/travis/activat.ion'` must be removed from .travis.yml in any case.

**Comment by achimmihca** [comment]
I did not yet create a PR with the new config also because I do not want to use a Unity license that has been activated by my private username and password.
Do you have an UltraStar specific mail account connected with Unity? If not then I would create one first, for example ultrastar.buildbot@gmail.com or ultrastar.play.buildbot@gmail.com

**Comment by basisbit** [comment]
yes, I tried to put the unity license file content into the TravisCI secrets area, but the WebUI of TravisCI had (has?) bugs so that linebreaks as well as escaped characters for linebreaks were not being saved and thus eventually I got too annoyed and just uploaded the file. Another suggested approach was to encrypt the file, add that to the git repo and store the secret for decrypting it in TravisCI secret config.
Anyways, I used my "private" unity license for that. The worst that could happen is that someone else uses the private license file to build their unity app - so I see no real world problem there, except that Unity could cancel the account for not complying with their license somehow (I allow others to use "my" license, so they can use Unity based build bot without having agreed to Unity license). Feel free to use a throwaway-email for creating that license file if you want to 😄
